### ISSUE TITLE
Batching from two string_input_producers does that don't preserve order.

### ISSUE THREAD

**Issue by panmari** [tensorflow/tensorflow#410]
I'm trying to read in multiple files (one file consists of the input, the other is the label). I assembled a pipeline using `tf.train.string_input_producer`, some custom decoder and `tf.train.batch` to put them together to batches. Unfortunately, the order does somehow get mixed up in this process. I extracted a minimal example that uses randomly generated strings instead of filenames (and also skips the decoding step), [see my gist here](https://gist.github.com/panmari/6bf13e06c21493d08c75)

In this example, all shuffling was disabled. But the error also occurs if I enabled shuffle and set the same seed for both string input producers. 

Am I using these classes not as intended? Or is there an actual bug?

**Comment by panmari** [comment]
I just realized that my test succeeds if I set `num_threads=1` in the `tf.train.batch` call. So the strange behavior seems to originate from there.

**Comment by josh11b** [comment]
I'm not sure exactly how you are trying to read the two files in parallel.  Doing that correctly can be a bit awkward, especially if you want to use multiple threads.  What reader(s) are you using?  Can you show the code you are using for your input pipeline?  The most robust approach would be a custom reader.

**Comment by panmari** [comment]
Well, one of them is my 'label image', the other the 'input image'. I don't have access to the code right now, but it's quite similar to the gist I posted in my initial post, just with some png.decode thrown inbetween.

**Comment by girving** [comment]
Looks like this fell through the cracks.  @panmari: Is this still an issue, and do you have code if so?  For better or worse, tensorflow exposes a lot of the complexity of asynchronous programming to users, so it may be that this is intended behavior.

**Comment by panmari** [comment]
I ended up going a completely different route, since the non-sequential nature of reading (rather small) png files this way ended up being the bottle neck of my learning pipeline. Starting from [this example](https://github.com/tensorflow/tensorflow/blob/r0.7/tensorflow/examples/how_tos/reading_data/convert_to_records.py), I wrote a small script that dumps multiple images into one `example` under different keys. That way, the two images are always read by the same reader, so I don't have to care about concurrency issues anymore. And since I dump 2000 examples into one file (which can be read sequentially), this way of reading data proved to be much faster (at least on traditional hard drives). 

I'll close this issue, since I feel this is the proper way to tackle this problem, not the one outlined in the original issue.

**Comment by CJxD** [comment]
I'm also having this issue. I might also have to start using TFRecords instead, but it would be nice to just have a way to ues two string_input_producers in conjunction.

**Comment by panmari** [comment]
I also went with TFRecords in the end. I simply joined the files in when
generating the TFRecords.

**Comment by mrry** [comment]
@CJxD It's possible that the new `tf.contrib.data` module could solve your problem. It's available in TensorFlow 1.2rc0, and there is some preliminary documentation here:

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/data/README.md

Assuming you have two text files, line delimited, I think the solution would look something like this:

```python
image_filenames = tf.contrib.data.TextLineDataset("image_index.txt")
images = image_filenames.map(lambda filename: tf.image.decode_image(tf.read_file(filename)))
output_target_strings = tf.contrib.data.TextLineDataset("output_targets.txt")
output_target_values = output_target_strings.map(lambda target: ...)  # Insert type conversion code here.

combined = tf.contrib.data.Dataset.zip((images, output_targets)).repeat().batch(BATCH_SIZE)
iterator = combined.make_one_shot_iterator()
next_element = iterator.get_next()
# ...
```

Note that building everything into a TFRecord file would have better I/O characteristics as @panmari notes, but this might work well for initial experimentation.
